### DISCUSSION TITLE
Signing issue with .p12 certificate from 2014 (Android)

### DISCUSSION THREAD

**Discussion by Marcophono2** [airsdk/Adobe-Runtime-Support#2457] [Q&A]
Hello!

I know that this is not a new issue and that it was intensively discussed with working solutions at the end but in my case I didn't find a working one.
I have an Android app in Google Play since 2014. Until last year I was working with Flash Builder on Windows. Bad solution, I know, but working until last year if regarding some things like switching back to Java 1.8.0_112 if exporting the apk. But now, with aab that does not work anymore. 

I am using the adt command line tool

`adt -package -target aab -storetype pkcs12 -keystore Certificates_Marc.p12 -storepass bla Marcophono.aab [...files...] -extdir ANEsMP -resdir res2`

resulting in

`Warning: failure in signing Android aab`

From the adt.log

```
Gradle build where ADT is using JRE version 11.0.15.1
but where Gradle will be using the JDK from C:/Java/jdk-11.0.15.1\bin\java.exe
Gradle build command:  C:/Java/jdk-11.0.15.1\bin\java.exe -Xmx1024m -Dorg.gradle.appname=gradlew -Dorg.gradle.jvmargs=-Xmx1024m -Duser.dir="C:\Temp\22b85c45-fa6e-4c59-a0ac-1381bb265a8e" -cp "C:\Temp\22b85c45-fa6e-4c59-a0ac-1381bb265a8e\gradle\wrapper\gradle-wrapper.jar";"C:\Progs\Adobe Flash Builder 4.7 (64 Bit)\eclipse\plugins\com.adobe.flash.compiler_4.7.0.349722\AIRSDK\lib\android\bin\gradle-cli.jar" org.gradle.wrapper.GradleWrapperMain :app:bundleRelease
com.adobe.argv.UsageError: could not load keystore file (password may be incorrect)
```
Belive me, the password was correct! 😄 

A second problem, very probably depending on the same reason, in the new aab Google Play signing procedure.
The pepk.jar tool Google offers to generate an uploading certificate does not like my old .p12 certificate:

```
jdk-11.0.15.1\bin\java -jar pepk.jar --keystore=Certificates_Marc.p12 --alias=1 --output=C:\0MarcophonoApp\certs_APP\Android\MPDE_Anfertigung_bundle_cert\enc.zip --encryptionkey=eb10fe8f7c7c9df715022017b00c6471f8ba8170b13049a11e6c09ffe3056a104a3bbe4ac5a955f4ba4fe93fc8cef27558a3eb9d2a529a2092761fb833b656cd48b9de6a
Enter password for store 'Certificates_Marc.p12':
Error: Unable to export or encrypt the private key
java.io.IOException: keystore password was incorrect
        at java.base/sun.security.pkcs12.PKCS12KeyStore.engineLoad(PKCS12KeyStore.java:2116)
        at java.base/sun.security.util.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:243)
        at java.base/java.security.KeyStore.load(KeyStore.java:1479)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.KeystoreHelper.loadKeystore(KeystoreHelper.java:53)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.KeystoreHelper.getKeystore(KeystoreHelper.java:39)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.ExportEncryptedPrivateKeyTool.run(ExportEncryptedPrivateKeyTool.java:207)
        at com.google.wireless.android.vending.developer.signing.tools.extern.export.ExportEncryptedPrivateKeyTool.main(ExportEncryptedPrivateKeyTool.java:165)
Caused by: java.security.UnrecoverableKeyException: failed to decrypt safe contents entry: java.security.cert.CertificateException: Unable to initialize, java.io.IOException: DerInputStream.getLength(): Redundant length bytes found
        ... 7 more
```
I also tried to export a jks from the .p12 and use that one for the pepk tool but got the same result.

If I try to use an older version of Java to use the pepk tool I am told that the supported class is too low (to run pepk). 

All together I think that I have to "update" my old .p12 file to a newer version. Is that correct? And if, how do I do that? And if not, what is the workaround?

AIRSDK 50.1.1.2
targetSdkVersion="33"
Android Studio set up with Android API 33 and Build Tools 33.0.1

I would be very happy to get a solution.

Best
Marc

**Comment by ajwfrost** [comment]
This sounds like you've got a certificate that's using a 1024-bit key..

> If I try to use an older version of Java to use the pepk tool I am told that the supported class is too low (to run pepk).

Which version did you use? See #1997 as it looks like (at some point) Java 1.8.0_122 was an option for this..

**Comment by Marcophono2** [comment]
Thank you, @ajwfrost ! 'I tried Jave 1.8.0_112. Aside from generating the encrypted file with pepk, I also still have to sign my aab with the old .p12 certificate, or not? And that also doesn't work. I tried to set back my Java version just for signing my aabwith my old .p12 file to 1.8.0_112 in my Windows path, also in my Android Studio but that also throws errors when I use adt via command line.

**Comment by marchbold** [comment]
Hi, 

When using an old certificate we suggest moving to google's App Signing, where you upload your old certificate and let Google handle signing releases using it. This is what you are doing with the pepk tool, you are creating an encrypted file that you send to Google to use for signing. 

Once you have done that you create a new "upload certificate" and use this to sign your aab. This new certificate you will be able to use with Java 11 so once you have used the pepk tool and sent Google your old certificate you can move to Java 11 for all your AIR/Android development. I'd suggest still keeping a copy of your release certificate but you shouldn't need to use it anymore. 

Some notes on the upload certificate and a guide to the whole process can be found here: 

https://airsdk.dev/docs/tutorials/platform/android/packaging-android-app-bundles#upload-certificate

@ajwfrost Just note it's not java `1.8.0_122` but `1.8.0_112` that still supports the 1024 bit certificates.

One last note, any services that you use that verify the signature of your app will need to be updated to also use the upload certificate signature, otherwise they will fail during testing, (they will still work at release however). This includes services like Google APIs (eg Play Games) and the Facebook SDK.

**Reply by marchbold** [reply]
Also just quickly on the errors:

These are due to using a version of Java > 1.8.0_112 and using an older 1024 certificate:

- `could not load keystore file (password may be incorrect)`
- `java.io.IOException: keystore password was incorrect`



The newer pepk tool doesn't support 1.8.0_112 which results in the other error:

- `supported class is too low (to run pepk).`

Use the version I linked: https://github.com/airsdk/Adobe-Runtime-Support/discussions/1997#discussioncomment-3059995

**Comment by Marcophono2** [comment]
Thank you very much, @marchbold ! I am sure this informations will help! I cannot test it right away because I used a wrong .p12 file for generate the encryted code which to enter in Google Play to get the .der signature code. I requested a setting back which was initiated meanwhile but takes up to 48 hours to become active.
My misconception was to think that I still have to sign my aab with the .p12 file and that I will need the .der file later at any time while the aab uploading process. 😄 
It was all very confusing for me, also because that other "wrong" .p12 file obviously contains the same .keystore but has other fingerprints though. I think I tried to put the keystore file into a newer .p12 file version a few years ago when the first time the problem came up with the apk signing since there was a certain Java version number reached. But it seems that I made something wrong there. My last app update happened 1.5 years ago and I cannot remember what tricks I used to sign it. Probably just temporary downgrading the Java version. But now that's not enought anymore (aab). But your informations will help me to bypass it.

Best regards, and thanks again to @ajwfrost and @marchbold !
Marc

**Reply by marchbold** [reply]
Good luck :) Let me know how you go!
